app/src/androidMain/kotlin/de/westnordost/streetcomplete/quests/level/AddLevelThing.kt

```diff
@@ -0,0 +1,112 @@
+import de.westnordost.streetcomplete.osm.isThing
+import de.westnordost.streetcomplete.util.math.contains
+import de.westnordost.streetcomplete.util.math.isInMultipolygon
+
+class AddLevelThing : OsmElementQuestType<String>, AndroidQuest {
+
+    /* including any kind of public transport station because even really large bus stations feel
+     * like small airport terminals, like Mo Chit 2 in Bangkok*/
+    private val mallFilter by lazy { """
+        ways, relations with
+         shop ~ department_store|mall
+         or aeroway = terminal
+         or railway = station
+         or amenity = bus_station
+         or public_transport = station
+    """.toElementFilterExpression() }
+
+    private val thingsWithLevelFilter by lazy { """
+        nodes, ways, relations with level
+    """.toElementFilterExpression() }
+
+    override val changesetComment = "Determine on which level things are in a building"
+    override val wikiLink = "Key:level"
+    override val icon = R.drawable.ic_quest_level_thing
+    /* disabled because in a mall with multiple levels, if there are nodes with no level defined,
+     * it really makes no sense to tag something as vacant if the level is not known. Instead, if
+     * the user cannot find the thing on any level in the mall, delete the element completely. */
+    override val isReplacePlaceEnabled = false
+    override val isDeleteElementEnabled = true
+    override val achievements = listOf(CITIZEN)
+
+    override fun getTitle(tags: Map<String, String>) = R.string.quest_level_thing_title
+
+    override fun getApplicableElements(mapData: MapDataWithGeometry): Iterable<Element> {
+        // get geometry of all malls in the area
+        val mallGeometries = mapData
+            .filter { mallFilter.matches(it) }
+            .mapNotNull { mapData.getGeometry(it.type, it.id) as? ElementPolygonsGeometry }
+        if (mallGeometries.isEmpty()) return emptyList()
+
+        // get all elements that have level tagged
+        val thingsWithLevel = mapData.filter { thingsWithLevelFilter.matches(it) }
+        if (thingsWithLevel.isEmpty()) return emptyList()
+
+        // with this, find malls that contain elements that have different levels tagged
+        val multiLevelMallGeometries = mallGeometries.filter { mallGeometry ->
+            var level: String? = null
+            for (element in thingsWithLevel) {
+                val pos = mapData.getGeometry(element.type, element.id)?.center ?: continue
+                if (!mallGeometry.bounds.contains(pos)) continue
+                if (!pos.isInMultipolygon(mallGeometry.polygons)) continue
+
+                if (element.tags.containsKey("level")) {
+                    if (level != null) {
+                        if (level != element.tags["level"]) return@filter true
+                    } else {
+                        level = element.tags["level"]
+                    }
+                }
+            }
+            return@filter false
+        }
```

Comment:
It would make sense to extract this into a helper function in LevelUtils.kt or similar, named e.g. `getMultiLevelMallGeometries(mapData: MapDataWithGeometry): List<ElementPolygonsGeometry>` to avoid duplication.

Or something. Maybe there is another function that could be extracted to remove some duplication.